src/uct/tcp/tcp_ep.c

```diff
@@ -65,14 +65,185 @@ static inline int uct_tcp_ep_ctx_buf_need_progress(uct_tcp_ep_ctx_t *ctx)
     return ctx->offset < ctx->length;
 }
 
+static inline void uct_tcp_ep_ctx_move(uct_tcp_ep_ctx_t *to_ctx,
+                                       uct_tcp_ep_ctx_t *from_ctx)
+{
+    if (uct_tcp_ep_ctx_buf_need_progress(from_ctx)) {
+        memcpy(to_ctx, from_ctx, sizeof(*to_ctx));
+        memset(from_ctx, 0, sizeof(*from_ctx));
+    }
+}
+
+static unsigned
+uct_tcp_ep_create_socket_and_connect_progress(void *arg)
+{
+    uct_tcp_ep_t *ep       = (uct_tcp_ep_t*)arg;
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    ucs_status_t status;
+
+    status = ucs_socket_create(AF_INET, SOCK_STREAM, &ep->fd);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    status = uct_tcp_iface_set_sockopt(iface, ep->fd);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    uct_tcp_cm_conn_start(ep);
+
+out:
+    return 1;
+
+err:
+    uct_tcp_ep_set_failed(ep);
+    goto out;
+}
+
+static ucs_status_t
+uct_tcp_ep_create_socket_and_connect(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface   = ucs_derived_of(ep->super.super.iface,
+                                              uct_tcp_iface_t);
+    uct_worker_cb_id_t cb_id = UCS_CALLBACKQ_ID_NULL;
+
+
+    if (ep->conn_retries++ > iface->config.max_conn_retries) {
+        ucs_error("tcp_ep %p: reached maximum number of connection retries "
+                  "(%u)", ep, iface->config.max_conn_retries);
+        uct_tcp_ep_set_failed(ep);
+        return UCS_ERR_TIMED_OUT;
+    }
+
+    uct_worker_progress_register_safe(&iface->super.worker->super,
+                                      uct_tcp_ep_create_socket_and_connect_progress,
+                                      ep, UCS_CALLBACKQ_FLAG_ONESHOT, &cb_id);
+
+    return UCS_OK;
+}
+
+int uct_tcp_ep_is_unconnected(uct_tcp_ep_t *ep)
+{
+    return (ep->conn_state == UCT_TCP_EP_CONN_STATE_CLOSED) &&
+           (ep->conn_retries == 0);
+}
+
+int uct_tcp_ep_is_connect_in_progress(uct_tcp_ep_t *ep)
+{
+    return (ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING) &&
+           (ep->conn_retries == 0);
+}
+
+static unsigned uct_tcp_ep_connect_progress(void *arg)
+{
+    uct_tcp_ep_t *ep       = (uct_tcp_ep_t*)arg;
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    uct_tcp_ep_t *peer_ep;
+
+    ucs_assert(uct_tcp_ep_is_connect_in_progress(ep));
+
+    if (uct_tcp_iface_is_self_addr(iface, &ep->peer_addr) ||
+        (peer_ep = uct_tcp_cm_get_ep(iface, &ep->peer_addr, ep->peer_conn_sn,
+                                     UCT_TCP_EP_FLAG_CTX_TYPE_RX)) == NULL) {
+        ucs_assert(!(ep->flags & UCT_TCP_EP_FLAG_CTX_TYPE_TX));
```

Comment:
maybe move this assert to `uct_tcp_ep_add_ctx_cap`?

Reply:
another one is on line 189
ok

Reply:
yes, I see
unfortunately, couldn't combine them, since sending REQ expect that no TX set yet